File: web/blocks/block-style.dart
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

part of NetTango;

class BlockStyle {
  static final String DEFAULT_STARTER_COLOR   = "#bb5555";
  static final String DEFAULT_CONTAINER_COLOR = "#8899aa";
  static final String DEFAULT_COMMAND_COLOR   = "#9977aa";
  static final String DEFAULT_TEXT_COLOR      = "#ffffff";
  static final String DEFAULT_BORDER_COLOR    = "#ffffff";
  static final String DEFAULT_FONT_FAMILY     = "Poppins, sans-serif";

  static BlockStyle DEFAULT_STARTER_STYLE   = new BlockStyle() .. blockColor = BlockStyle.DEFAULT_STARTER_COLOR;
  static BlockStyle DEFAULT_CONTAINER_STYLE = new BlockStyle() .. blockColor = BlockStyle.DEFAULT_CONTAINER_COLOR;
  static BlockStyle DEFAULT_COMMAND_STYLE   = new BlockStyle() .. blockColor = BlockStyle.DEFAULT_COMMAND_COLOR;

  String blockColor  = DEFAULT_COMMAND_COLOR;
  String textColor   = DEFAULT_TEXT_COLOR;
  String borderColor = DEFAULT_BORDER_COLOR;
  String fontWeight  = "";
  String fontSize    = "";
  String fontFace    = "";

  String get font {
    final weight = fontWeight == "" ? "" : "font-weight: $fontWeight;";
    final size   = fontSize   == "" ? "" : "font-size: $fontSize;";
    final face = "font-family: ${ fontFace == "" ? DEFAULT_FONT_FAMILY : fontFace };";
    return "$weight $size $face".trim();
  }

  String get cssRule =>
    "color: $textColor; border-color: $borderColor; $font".trim();

  appendToSheet(CssStyleSheet sheet, String blockClass) {
    sheet.insertRule(".$blockClass-color { background-color: $blockColor; }", 0);
    sheet.insertRule(".$blockClass-attribute { color: $blockColor; background-color: $textColor; }", 0);
    sheet.insertRule(".$blockClass { $cssRule }", 0);
  }

}

File: src/blocks/block-style.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import { BlockStyle } from "../types/types"
import { ObjectUtils } from "../utils/object-utils"

class BlockStyleUI {

  static readonly DEFAULT_STARTER_COLOR   = "#bb5555"
  static readonly DEFAULT_CONTAINER_COLOR = "#8899aa"
  static readonly DEFAULT_COMMAND_COLOR   = "#9977aa"
  static readonly DEFAULT_TEXT_COLOR      = "#ffffff"
  static readonly DEFAULT_BORDER_COLOR    = "#ffffff"
  static readonly DEFAULT_FONT_FAMILY     = "Poppins, sans-serif"

  static readonly DEFAULT_COMMAND_STYLE: BlockStyle = {
    blockColor:  BlockStyleUI.DEFAULT_COMMAND_COLOR
  , textColor:   BlockStyleUI.DEFAULT_TEXT_COLOR
  , borderColor: BlockStyleUI.DEFAULT_BORDER_COLOR
  , fontWeight:  ""
  , fontSize:    ""
  , fontFace:    ""
  }
  static readonly DEFAULT_CONTAINER_STYLE = ObjectUtils.clone(BlockStyleUI.DEFAULT_COMMAND_STYLE, { blockColor: BlockStyleUI.DEFAULT_CONTAINER_COLOR })
  static readonly DEFAULT_STARTER_STYLE   = ObjectUtils.clone(BlockStyleUI.DEFAULT_COMMAND_STYLE, { blockColor: BlockStyleUI.DEFAULT_STARTER_COLOR })

  readonly bs: BlockStyle

  constructor(bs: BlockStyle) {
    this.bs = bs
  }

  get font() {
    const weight = this.bs.fontWeight === "" ? "" : `font-weight: ${this.bs.fontWeight};`
    const size   = this.bs.fontSize   === "" ? "" : `font-size: ${this.bs.fontSize};`
    const face = `font-family: ${ this.bs.fontFace === "" ? BlockStyleUI.DEFAULT_FONT_FAMILY : this.bs.fontFace };`
    return `${weight} ${size} ${face}`.trim()
  }

  get cssRule() {
    return `color: ${this.bs.textColor}; border-color: ${this.bs.borderColor}; ${this.font}`.trim()
  }

  appendToSheet(sheet: CSSStyleSheet, blockClass: string) {
    sheet.insertRule(`.${blockClass}-color { background-color: ${this.bs.blockColor}; }`, 0)
    sheet.insertRule(`.${blockClass}-attribute { color: ${this.bs.blockColor}; background-color: ${this.bs.textColor}; }`, 0)
    sheet.insertRule(`.${blockClass} { ${this.cssRule} }`, 0)
  }

}

export { BlockStyleUI }
